File: core/reasoning/semantic_constraint_engine.py
from __future__ import annotations

from typing import Any, Dict


def apply_semantic_constraints(bundle: Dict[str, Any], max_depth: int = 20) -> Dict[str, Any]:
    lineage = bundle.get("lineage", {}) or {}
    depth = lineage.get("depth", 0) if isinstance(lineage, dict) else 0
    bounded = depth <= max_depth
    return {**bundle, "constraints": {"max_depth": max_depth, "satisfied": bounded}}

File: src/reasoning/semanticConstraintEngine.ts
/**
 * Converted from Python: core/reasoning/semantic_constraint_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function applySemanticConstraints(bundle: any, max_depth: any = 20): any {
  var lineage: any = py.or2(py.get(bundle, "lineage", {}), () => ({}));
  var depth: any = (((lineage !== null && typeof lineage === "object" && !Array.isArray(lineage) && !(lineage instanceof Set) && !(lineage instanceof Map))) ? py.get(lineage, "depth", 0) : 0);
  var bounded: any = py.le(depth, max_depth);
  return {...(bundle), "constraints": {"max_depth": max_depth, "satisfied": bounded}};
}
